code/modules/projectiles/guns/energy/laser.dm

```diff
@@ -126,39 +126,27 @@
 	weapon_weight = WEAPON_HEAVY
 	origin_tech = "combat=6;magnets=6;powerstorage=4"
 	ammo_type = list(/obj/item/ammo_casing/energy/laser/sniper, /obj/item/ammo_casing/energy/laser/sniper/pierce)
-	zoomable = TRUE
-	zoom_amt = 7
 	shaded_charge = TRUE
-	/// Is the scope fully online or not?
-	var/scope_active = FALSE
-	var/stored_dir
 	execution_speed = 8 SECONDS
 
-/obj/item/gun/energy/lwap/zoom(mob/living/user, forced_zoom)
+/obj/item/gun/energy/lwap/Initialize(mapload)
 	. = ..()
-	if(!ishuman(user))
-		return
-	var/mob/living/carbon/human/H = user
-	stored_dir = H.dir
-	if(scope_active && !zoomed)
-		select_fire(H)
-		H.remove_status_effect(STATUS_EFFECT_LWAPSCOPE)
-		scope_active = FALSE
-		return
-	if(zoomed && do_after(user, 3 SECONDS, target = src))
-		if(zoomed && !scope_active) //We check after to be sure.
-			scope_active = TRUE
-			to_chat(user, "<b><span class='robot'>SCOPE_CREEPER_[rand(1, 9999)] Online.</span></b>")
-			select_fire(H)
-			H.apply_status_effect(STATUS_EFFECT_LWAPSCOPE, stored_dir)
-		return
-	if(zoomed)
-		zoom(user, FALSE) //Moved while scope was booting, so we unzoom
+	AddComponent(/datum/component/scope, range_modifier = 2, time_to_scope = 3 SECONDS)
+
+/obj/item/gun/energy/lwap/on_scope_success(mob/living/user)
+	to_chat(user, "<b><span class='robot'>SCOPE_CREEPER_[rand(1, 9999)] Online.</span></b>")
+	select_fire(user)
+	user.apply_status_effect(STATUS_EFFECT_LWAPSCOPE)
+
+/obj/item/gun/energy/lwap/on_scope_end(mob/living/user)
+	select_fire(user)
+	user.remove_status_effect(STATUS_EFFECT_LWAPSCOPE)
 
 /obj/item/gun/energy/lwap/on_mob_move(dir, mob/user)
-	if(scope_active)
+	if(HAS_TRAIT(user, TRAIT_SCOPED) && istype(chambered, /obj/item/ammo_casing/energy/laser/sniper/pierce)) //This just makes sure it's our zoom
 		to_chat(user, "<span class='warning'>[src]'s scope is overloaded by movement and shuts down!</span>")
-		zoom(user, FALSE)
+		var/datum/component/scope/ourscope = GetComponent(/datum/component/scope)
```

Comment:
Isn't this already solved by the component itself?

Reply:
The component does not stop scoping on moving.

Reply:
This has been changed to be handled by the gun